Add new analyses and bring code more in line with JASP standards

jansim opened the pull request:
This PR jointly adds the new analyses / features from [LSLindeloo/jaspPower/powerModule](https://github.com/LSLindeloo/jaspPower/tree/powerModule) by @LSLindeloo and the rewrtie of large parts of the codebase to be more in line with JASP standards [jansim/jaspPower/#9](https://github.com/jansim/jaspPower/pull/9).

The JASP-style rewrite mainly came with a transition from an R6 based object oriented style to a more functional style, following JASP function naming conventions e.g. prefixing non-exported functions with a `.`.

This PR also includes an upgrade and matching fixes to use the new unit test specification.

@LSLindeloo not sure whether you may want to elaborate on the new analyses / features here?

Kucharssim commented on R/pwr_1p.R:
```diff
@@ -1,702 +1,1026 @@
+
+  if (calc == "sampleSize") {
+    str <- gettextf(
+      "We would need %1$s to reliably (with probability greater than or equal to %2$s) detect an effect size of <i>|h|%3$s</i>%4$s, assuming a %5$s criterion for detection that allows for a maximum Type I error rate of <i>\u03B1=</i>%6$s.",
```
html tahs and unicode symbols should be passed into gettextf as arguments; I assume we don't want the translators with changing e.g., `<i>` to `<br>`!

Kucharssim commented on R/pwr_1p.R:
```diff
@@ -1,702 +1,1026 @@
+      alternative = alt
+    )$p[1]
+  }))
+  if (inherits(probs_es, "try-error")) {
```
for these checks you could use `jaspBase::isTryError`

Kucharssim commented on R/pwr_1var.R:
```diff
@@ -1,639 +1,1033 @@
+
+  ## Compute results
+  results <- try(.computeTest1Var(jaspResults, options, stats))
+  if (inherits(results, "try-error")) {
+    .quitAnalysis(gettext("Unable to compute the power results. Try to enter less extreme values for the input parameters."))
+  }
```
This seems to be on top of each of the analyses, could you move this into a function and just reuse the function?

Kucharssim commented on R/pwr_2p.R:
```diff
@@ -1,748 +1,1048 @@
+
+  .initPowerTabTest2P(jaspResults, options, results, stats)
+
+  if (options$text) {
+    .initPowerESTabTest2P(jaspResults, options, results, stats)
+  }
+
+  ## Populate tables and plots
+  if (options$text) {
+    .populateIntro(jaspResults, options)
+  }
+
+  if (options$powerContour) {
+    .preparePowerContourTest2P(jaspResults, options, results, stats)
+    if (options$text) {
+      .populateContourTextTest2P(jaspResults, options, results, stats)
```
These parts are also a bit repetetive...

For example, instead of having in each analysis

```r
if(options[["text"]])
  .populateIntro(jaspResults, options)
```

You could just define `.populateIntro` with
```
.populateIntro <- function(jaspResults, options) {
  if(!options[["text"]]) return()

  # rest of the original function
}
```

Then you don't have to do the check every time you call this function and as a result the analysis functions become a bit more readable.

jansim replied:
Ah, yeah, this is the part where I struggled a bit to find a nice way of porting this over from the object oriented programming style (which solves this more elegantly).

Will update this for any common functions, I think for the analysis specific functions it's probably more readable this way? And it would be duplicated either way 🤔

Kucharssim reviewed (changes requested):
Looks very nice. I have not tested very thoroughly in JASP, but it seems to work ok for majority of "reasonable" inputs. Pushing it to less sensible regions leads to informative error messages, which is good.

I provided some code comments regarding some style and code duplication - these would be nice to polish but it's not absolutely necessary. There are some issues with the use of `gettextf` which we definitely need to fix before merging though.

shun2wang commented on inst/qml/Power.qml:
```diff
@@ -56,124 +60,164 @@ Form
+        {
+          id: direction
+          name: "effectDirection"
+          label: qsTr("")
```
Hi, we shoud not mark a empty string as to be translated:-) and [here](https://github.com/jasp-stats/jaspModuleTemplate/pull/8) we will have a test to check this.

jansim commented on R/commonPower.R:
```diff
@@ -0,0 +1,695 @@
+  es <- options$effectSize
+  alpha <- options$alpha
+
+  if (pow >= 1) .quitAnalysis(gettext("Power must be less than 1."))
```
I think the max value is already set in Power.qml:159 , so this is just an additional check. I think it makes sense to leave it in, given that there's also the idea floating around of supporting the use of jasp modules within R (or am I mistaken here? 🤔)

Kucharssim replied:
> supporting the use of jasp modules within R (or am I mistaken here? 🤔)

No you are correct, but the idea is that if the user calls a JASP function in R, the options are sent to QML for validation first; the input in R will go through the same checks as in the JASP GUI - to exactly avoid having to implement these checks both in R and QML.

jansim replied:
Oh wow, what a fancy solution! At some point I'll have to take a look how you parsed the QML and all! 👀 

I'll remove the line then 👍

jansim commented on R/commonPower.R:
```diff
@@ -0,0 +1,695 @@
+  )
+
+  if (calc == "sampleSize") {
+    mid_sentence <- gettextf(
```
I think I cought every one of these now, thanks!

jansim commented on R/commonPower.R:
```diff
@@ -0,0 +1,695 @@
+    " when using <i>%s</i>.", test_names[[options$test]]
+  )
+
+  if (calc == "sampleSize") {
```
Good one, I've always been a bit hesitant about using `switch` in R due to its weird implementation as a function. This is a really good use case of it though so will change it, ty! 👍️

jansim commented on R/commonPowerPlotting.R:
```diff
@@ -0,0 +1,315 @@
+    )
+  }
+
+  # Checking whether geom_contour_filled is availaible
+  # (as it is e.g. not yet in the ggplot2 version of JASP 0.14.1)
+  # TODO: Remove this check in the future
```
Good catch! ✅

jansim commented on R/commonPower.R:
```diff
@@ -0,0 +1,695 @@
+}
+
+# Workaround for failure of pwr::.pwrT2NTest
+# with large effect sizes - optimization here is
```
This is actually coming from jpower, not 100% sure what Richard changed here and whether it's worth digging into? I updated the comment for now to clearly reference the source of the code 👍️

Kucharssim replied:
ah ok, then this is ok

jansim commented:
Thank you for these very detailed and super helpful tips!! 🙏🙏

I did a first pass on this, adressing most of the comments. Feel free to re-open comments, if you disagree with me marking it as resolved (this was also largely for my own tracking of what's addressed).

I managed to deal with about 50% of the `gettextf` related issues, unfortunately these take a huge amount of time since they are scattered throughout the whole codebase and can't be solved with find / replace (at least that I know of). @LSLindeloo do you think you could maybe help here in fixing the last ones of these? This is mainly about getting rid of any HTML in translated strings and replacing unordered `%s` with `%1$s`, `%2%s`, ... for strings with multiple parameters. Either way it's probably good to have a second pair of eyes on this since it's a lot of just shuffling things around.

Kucharssim commented on R/commonPowerPlotting.R:
```diff
@@ -0,0 +1,315 @@
+# ==== Default Plotting Configurations ====
+
+# Common theme to be used for all plots
+.pwrPlotTheme <- function() {
```
> I think themeJasp itself actually deprecated? 

You are right!

In this case I would compare to

```r
plotObject +
  jaspGraphs::geom_rangeframe() +
  jaspGraphs::themeJaspRaw(legend.position = "right")
```

The idea is to have a uniform style of plots across JASP, so it makes sense to stick to that. Unless there is a reason why the default JASP style does not work for the particular plot (that does not seem the case here but you can convince me otherwise).

> If this is important, I guess we could just quickly change it over as this is handled in the function anyways. 🤷‍♂️

Yes, it's nice that you wrapped it in a separate function!

An alternative is to let the user switch between different themes (this is for example done in Mixed Models module) if you think it's worth for people to try different looks. But usually it's not worth the effort.

jansim commented:
@Kucharssim I've updated the plots to use the jaspGraphs theme, however, this requires the snapshots to be updated and I can only do so for macOS (since some of the plots are platform dependent). Could you re-run the tests locally and upload the plots if you are based on either Windows or Linux? (Or do you know someone from the team who could do this?)

jansim commented:
Thank you, Don! No idea why it wasn't possible, but I merged the changes and all tests pass :tada:

Kucharssim commented on inst/qml/Power.qml:
```diff
@@ -56,124 +60,164 @@ Form
+          id: alpha
+          name: "alpha"
+          label: qsTr("α")
+          min: 0
+          defaultValue: 0.05
```
This option needs `max`

Kucharssim reviewed (changes requested):
There are still some details to fix in the QML validation, otherwise there is only some gettexting to be done; I would recommend one more walk through the code to polish that out. Otherwise it looks great!

> I also tried to catch the last remaining occurences of HTML in gettext / gettextf, not sure whether there's an option of exporting all the strings to catch any remaining bits that might've slipped through?

Well we have a GitHub action that will generate the to-be-translated strings. But usually that runs after we merge a PR. Of course we could merge it as is and make another PR to clean up later, but it would be really better to do as much as we can now already. Unless @shun2wang or @vandenman have any other idea how to make this process more convenient.

> @jansim I've opened https://github.com/jansim/jaspPower-fork-only/pull/1 which regenerates the test figures. For some reason I can't push to this PR directly, otherwise I would have done so.

It seems I was the only administrator of this repo; I just changed the permissions to jasp teams according to other modules.

shun2wang commented:
> I also tried to catch the last remaining occurences of HTML in `gettext` / `gettextf`, not sure whether there's an option of exporting all the strings to catch any remaining bits that might've slipped through?

Hi, [here](https://github.com/jasp-stats/jasp-actions/blob/dd3afbaa5162c1efd04f58db6c6e06639585bde5/translations/i18nCheck.R#L10) we have a action to catch all the strings to be translated as a dataframe. we didn't export them in the GitHub wrokflow. but we can do this if necessary.
You can also run this tool locally:

```r
library(potools)
rPotData <- potools::get_message_data(dir = ".", verbose = TRUE)
View(rPotData )
```
Then you can see all the strings with translations listed in the call column (this also contains additional` stop/warning...` calls, usually you should ignore them because JASP now only treats calls explicitly marked as `gettext/gettextf` as strings to be translated).

jansim commented:
Thank you for the detailed feedback again and that amazing snippet (@shun2wang)!

I've eradicated the last bits of HTML in translated strings and updated the QML to use the interval setting correctly 👍️

I used the following snippet to detect all messages containing HTML:

```R
library(potools)
library(tidyverse)
potools::get_message_data(dir = ".", verbose = TRUE) %>% 
  filter(
    str_detect(msgid, fixed("<b>")) |
    str_detect(msgid, fixed("<i>")) |
    str_detect(msgid, fixed("<p>"))
  ) %>% 
  View()
```

shun2wang commented on R/pwr_2var.R:
```diff
@@ -1,728 +1,1149 @@
+
+  esText <- c(
+    gettextf("1 %1$s %2$s %3$s  %4$s", sign1, "\u03C1", sign2, format(round(as.numeric(probs_es[1]), 3), nsmall = 3)),
+    gettextf("%1$s %2$s %3$s %4$s %5$s", format(round(as.numeric(probs_es[1]), 3), nsmall = 3), sign1, "\u03C1", sign2, format(round(as.numeric(probs_es[2]), 3), nsmall = 3)),
```
These seem also not need to be translated, because translator will do nothing for them. but it's ok if you don't have time to clean these.

jansim replied:
Good catch! I replaced gettextf with sprintf in these cases (couldn't be bothered to change it into paste, but maybe @LSLindeloo can update parts here if the code around it is being changed anyways at some point, AFAIK he did want to do some minor changes there)

jansim replied:
Also FYI if anyone runs into a similar problem in the future, I was able to easiliy find these strings by searching for the ones with a high ratio of `%` compared to overall text length 😄

Kucharssim commented:
Looks great! Thanks @jansim for the additional effort you had to put in and @shun2wang for a another pair of eyes in the review!
